### PR TITLE
First pass at implementing OpenROAD flow using SiliconCompiler

### PR DESCRIPTION
**Pull request by WRansohoff**
(Internal code review PR)

This change adds a general SiliconCompiler build flow to run the OpenROAD flow represented in this repository. It also includes platform targets for nangate45 and sky130hd.

While there are still some places where we could better integrate the ORFS environment variables with the SiliconCompiler JSON schema, this should suffice as a demonstration for implementing Makefile logic using the SiliconCompiler Python library.

We tried to maintain a similar syntax to the existing Make-based solution. To run an example design (currently only nangate45/sky130hd platforms are supported):

`python3 Makefile.py -DESIGN_CONFIG=designs/[platform]/[design]/config.mk`

These scripts have been tested on the set of designs which appear to be used in the ORFS CI tests:

* nangate45/
  * aes
  * black_parrot
  * bp_be_top
  * bp_fe_top
  * bp_multi_top
  * dynamic_node
  * ibex
  * jpeg
  * swerv
  * swerv_wrapper
  * tinyRocker
* sky130hd/
  * aes
  * chameleon
  * gcd
  * ibex
  * jpeg
  * microwatt
  * riscv32i

### CODE REVIEW

File: flow/Makefile.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        # Update 'config' dictionary with the substitutions used to generate it.
+        for k, v in repls.items():
+            config[k] = v
+
+    chip = siliconcompiler.Chip(design)
+    chip.set('option', 'scpath', scdir)
+
+    # Extract sources, and convert to absolute paths b/c sc commands run in a different build dir.
+    if 'SDC_FILE' in config.keys():
+        sdc_in = os.path.abspath(config.pop('SDC_FILE'))
```
Comment (nmoroze) [inline flow/Makefile.py:60]: Use find_files() when we read this out in target

File: flow/Makefile.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        config = parse_config_mk.parse(args['DESIGN_CONFIG'])
+        # Update 'config' dictionary with the substitutions used to generate it.
+        for k, v in repls.items():
+            config[k] = v
+
+    chip = siliconcompiler.Chip(design)
+    chip.set('option', 'scpath', scdir)
+
+    # Extract sources, and convert to absolute paths b/c sc commands run in a different build dir.
+    if 'SDC_FILE' in config.keys():
```
Comment (nmoroze) [inline flow/Makefile.py:59]: generalize logic for files

File: flow/Makefile.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                continue
+            taskdir = chip._getworkdir(step=step)
+            chip.set('tool', 'openroad', 'env', step, '0', 'RESULTS_DIR', os.path.join(taskdir, WORKDIR_NAME))
+            # TODO: decide what to do about objects directory.
+
+    # Build the design.
+    chip.run()
+
+    chip.summary()                                # print results summary
+    chip.show()                                   # show layout file
```
Comment (WRansohoff) [inline flow/Makefile.py]: Remove

File: flow/scripts/klayout.tcl [hunk 4]
```diff
@@ -10,11 +10,14 @@ if {[info exists ::env(SEAL_GDS)]} {
     set seal_gds ""
 }
 
+set GDSOAS_INPUT "$::env(GDSOAS_FILES) $::env(WRAPPED_GDSOAS)"
```
Comment (WRansohoff) [inline flow/scripts/klayout.tcl]: Call out modified files explicitly in the PR to the ORFS repo

File: flow/scripts/sc/examples/heartbeat/build.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
+import os
```
Comment (WRansohoff) [inline flow/scripts/sc/examples/heartbeat/build.py]: Delete heartbeat example, use basic designs (gcd, etc)

File: flow/scripts/sc/flows/orflow.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    chip.set('option', 'showtool', 'gds', 'klayout')
+
+    # Setup empty 'import' stage with a single-node OpenROAD task to run 'run_all.tcl'
+    # TODO: Use the 'import' step to parse config.mk file env vars.
+    # TODO: Split up individual TCL scripts into sc tasks?
+    flow = 'orflow'
+
+    flow_groups = {
+        'synthesis': [
+            ('or_yosys', 'openroad'), # (synthesis is done via OpenROAD TCL that calls yosys)
```
Comment (WRansohoff) [inline flow/scripts/sc/flows/orflow.py]: Use yosys tool instead of yosys.tcl run thru openroad

File: flow/scripts/sc/flows/orflow.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            ('or_fillcell', 'openroad'),
+        ],
+        'route': [
+            ('or_global_route', 'openroad'),
+            ('or_detail_route', 'openroad'),
+        ],
+        'finish': [
+            ('or_final_report', 'openroad'),
+            # Like yosys, KLayout GDS-streaming script is run through an OpenROAD tcl script
+            ('export', 'openroad')
```
Comment (WRansohoff) [inline flow/scripts/sc/flows/orflow.py]: Use klayout tool instead of klayout.tcl run thru openroad

File: flow/scripts/sc_apr.tcl [hunk 8]
```diff
@@ -0,0 +1,117 @@
+source sc_manifest.tcl
```
Comment (nmoroze) [inline flow/scripts/sc_apr.tcl]: move under sc directory

Comment (WRansohoff) [inline flow/scripts/sc_apr.tcl]: Move to scripts/sc/tools/...

File: flow/scripts/sc/libs/nangate45_orflow.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+    # filler
+    lib.add('asic', 'cells', 'filler', ["FILLCELL_X1",
+                                        "FILLCELL_X2",
+                                        "FILLCELL_X4",
+                                        "FILLCELL_X8",
+                                        "FILLCELL_X16",
+                                        "FILLCELL_X32"])
+
+    # Stupid small cells
```
Comment (WRansohoff) [inline flow/scripts/sc/libs/nangate45_orflow.py]: "Mark small cells as dontuse"

File: flow/scripts/sc/targets/nangate45_orflow.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    chip.set('asic', 'coremargin', 1.9)
+    # Set timing corners.
+    corner = 'typical'
+    chip.set('constraint','worst','libcorner', corner)
+    chip.set('constraint','worst','pexcorner', corner)
+    chip.set('constraint','worst','mode', 'func')
+    chip.set('constraint','worst','check', ['setup','hold'])
+
+    # TODO: Add an 'objects' directory in the task dirs? Use 'outputs/' for now.
+    env_vars = {
```
Comment (WRansohoff) [inline flow/scripts/sc/targets/nangate45_orflow.py:61]: Replace hardcoded env vars w/ schema values where possible.

File: flow/scripts/sc/targets/sky130hd_orflow.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    chip.set('asic', 'aspectratio', 1)
+    chip.set('asic', 'coremargin', 62.56)
+    corner = 'typical'
+    chip.set('constraint', 'worst', 'libcorner', corner)
+    chip.set('constraint', 'worst', 'pexcorner', corner)
+    chip.set('constraint', 'worst', 'mode', 'func')
+    chip.add('constraint', 'worst', 'check', ['setup','hold'])
+
+    # TODO: Add an 'objects' directory in the task dirs? Use 'outputs/' for now.
+    env_vars = {
```
Comment (WRansohoff) [inline flow/scripts/sc/targets/sky130hd_orflow.py:59]: Replace hardcoded values with schema values where possible.

File: flow/scripts/sc_apr.tcl (hunk 8 above)
Comment (WRansohoff) [inline flow/scripts/sc_apr.tcl]: More comments: purpose of file, description of each section and individual steps

File: flow/scripts/klayout.tcl (hunk 4 above)
Comment (WRansohoff) [reply flow/scripts/klayout.tcl]: No more modified files, after moving away from the possibly-deprecated klayout.tcl script.

**Comment by WRansohoff** [comment]
I added an ASAP7 target, and the 'mergeLib' step which appeared to be optional for the nangate45/sky130hd platforms. This lets us run the following asap7 CI tests, which all passed:

* asap7/aes
* asap7/ethmac
* asap7/gcd
* asap7/ibex
* asap7/jpeg
* asap7/sha3
* asap7/uart

So this build flow should support the entire Jenkins test suite used in the OpenROAD-flow-scripts repository, except for the sky130hs platform. We do not currently have a sky130hs target in the SiliconCompiler repository, but it may not be difficult to implement considering the similarity to sky130hd.

File: flow/Makefile.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
+import argparse
+import glob
+import os
+import shutil
+import sys
+
+import siliconcompiler
+
```
Comment (WRansohoff) [inline flow/Makefile.py:8]: Add comment pointing to scripts/sc/...

**Comment by WRansohoff** [comment]
I added the sky130hs target, and it looks like that platform's CI tests all pass:

* sky130hs/aes
* sky130hs/gcd
* sky130hs/ibex
* sky130hs/jpeg
* sky130hs/riscv32i

**Pull request merged**